docs/spawn3.man
.\" RCSID @(#)$Id: spawn3.man,v 1.2 2002/10/22 21:39:45 rk Exp $
.\" LIBDIR
.TH "SPAWN" "3rko" "2 Jun 1995"
.SH NAME
spawn3 \- spawn3 a child process with stdin, stdout, and stderr
stream communication.

.SH SYNOPSIS

.nf
#include <stdio.h>
#include "coexec.h"

int spawn3(FILE **childin, FILE **childout, FILE **childerr, char **argv);
.fi

.SH DESCRIPTION
.I spawn
sets up a child process and communications to the process via the
stdin, stdout, and stderr FILE streams.
.I argv
is a command vector, where argv[0] is the executable to run, and
argv[1],... are the commandline options.
Must have the last one argv[N] = NULL to terminate the vector.

.I spawn3
returns the child pid if successful, else < 0 an error.
Pass in the address to the FILE streams
.IR childin , childout ,
and
.I childerr
for communication to the spawned process.

.SH SEE ALSO
coexec(3rko),invoke(3rko),spawn(3rko),uvec(3rko),
execvp(2),fork(2),pipe(2),select(2),fclose(3),wait(2),waitpid(2),system(3)

.SH NOTES

If both stdout and stderr outputs need to be read, then the calling program
should
.I select
between the two file streams since the child process may be writing
to either at any time.

.nf
 /* stdall - collects and outputs the stdout,stderr, and exit status
  * takes all the options and just passes them on to a sub-process
  */
 
 #include <stdio.h>
 #include <sys/types.h>	/* waitpid,select */
 #include <sys/wait.h>	/* waitpid */
 #include <unistd.h>	/* select */
 #include <sys/time.h>	/* select */
 #include "coexec.h"	/* spawn3 */
 #include "uvec.h"
 
 #define MAXBUFF	512
 
 int main(int argc, char *argv[]) {
 	FILE	*childin,
 		*childout,
 		*childerr;
 	pid_t	 childpid;
 	int	status,
 		 retexit = 0;
 	char	 bufout[MAXBUFF],
 		 buferr[MAXBUFF];
 	fd_set	 cfds;
 	int	 maxfd,selval;
 	struct timeval tv;
 	uvec	*cerr;
 
 	if (argc < 2) return 1;
 
 	/* spawn the child process */
 	if ((childpid = spawn3(&childin,&childout, &childerr, argv+1)) < 0) {
 		fprintf(stderr,"%s failed to spawn3 '%s' ret = %d\n",
 			*argv, argv[1], childpid);
 		return 2;
 	}
 	/* close stdin to the child - don't need it here */
 	fclose(childin);
 
 	/* watch child stdin/stderr */
 	FD_ZERO(&cfds);
 	FD_SET(fileno(childout),&cfds);
 	FD_SET(fileno(childerr),&cfds);
 	tv.tv_sec = 0;
 	tv.tv_usec = 100;
 
 	/* capture stderr to an expandable array */
 	cerr = uvec_ctor(20);
 	maxfd = (fileno(childout) > fileno(childerr)
 		? fileno(childout) : fileno(childerr)) + 1;
 	printf(
 "== stdout ============================================================\n");
 
 	while ((selval = select(maxfd,&cfds,NULL,NULL,&tv))
       	> 0) {
 		if (FD_ISSET(fileno(childout),&cfds)) {
 			*bufout = '\0';
 			fgets(bufout, MAXBUFF, childout);
 			/* output stdout now */
 			if (*bufout) printf("%s", bufout);
 		} else if (FD_ISSET(fileno(childerr),&cfds)) {
 			*buferr = '\0';
 			fgets(buferr, MAXBUFF, childerr);
 			/* store stderr for later output */
 			uvec_push(cerr,buferr);
 		}
 		/* reset the select input */
 		FD_ZERO(&cfds);
 		if (!feof(childout)) FD_SET(fileno(childout),&cfds);
 		if (!feof(childerr)) FD_SET(fileno(childerr),&cfds);
 		tv.tv_sec = 0;
 		tv.tv_usec = 100;
 	}
 
 	fclose(childout);
 	fclose(childerr);
 
 	/* may still lose some stderr if there is lots of it */
 	argv = uvec_vector(cerr);
 	printf(
 "== stderr ============================================================\n");
 	while (*argv) {
 		printf("%s",*argv++);
 	}
 	uvec_dtor(&cerr);
 
 	printf(
 "== exit   ============================================================\n");
 	waitpid(childpid, &status, 0);
 	if (WIFEXITED(status)) retexit = WEXITSTATUS(status);
 	printf("exit = '%d'\n", retexit);
 	printf(
 "== end    ============================================================\n");
 	return retexit;
 }

.fi

After the output is read and an EOF is received, the calling program should
close the file stream and wait for the child process.

.SH DIAGNOSTICS

The errors are handled by rkoperror() (if RKOERROR is defined),
else it goes to stderr.

.SH BUGS
None discovered yet.

.SH AUTHOR
R.K.Owen,Ph.D.

.KEY WORDS


NEWS

librko-1.0.0	2002/10/21
* first release with a working configure, using automake/autoconf etc.
* added spawn3



coexec/spawn3.c
static const char RCSID[]="@(#)$Id: spawn3.c,v 1.2 2002/10/22 21:39:45 rk Exp $";
static const char AUTHOR[]="@(#)coexec 1.0 2002/02/28 R.K.Owen,Ph.D.";
/* coexec.c - contains all the co-executable routines
 */
/** ** Copyright *********************************************************** **
 ** 									     **
 ** Copyright 2002 by R.K.Owen,Ph.D.		                      	     **
 ** last known email: librko@kooz.sj.ca.us				     **
 **                   rk@owen.sj.ca.us					     **
 ** see LICENSE.LGPL, which must be provided, for details		     **
 ** 									     **
 ** ************************************************************************ **/
#define _POSIX_SOURCE

#include <stdio.h>	/* perror */
#include <stdlib.h>	/* _exit */
#include <string.h>	/* strcpy, strcat, etc */
#include <sys/types.h>
#include <unistd.h>	/* fork,exec,dup */
#include <signal.h>	/* kill */
#ifdef HAVE_CONFIG_H
#  include "config.h"
#endif
#include "coexec.h"
#ifdef RKOERROR
#  include "rkoerror.h"
#endif

/*----------------------------------------------------------------------*/
/* spawn3.c - routine for fork/exec/dup/pipe a child process and tie its
 *	input/output/stderr back to be written/read/read by the caller.
 *	returns the child pid (>0) if successful, else < 0 if error.
 */

int spawn3(FILE **childin, FILE **childout, FILE **childerr, char **argv) {
	int pipepath1[2];
	int pipepath2[2];
	int pipepath3[2];
	int parent_rfd, parent_wfd, parent_efd;
	int child_rfd,  child_wfd, child_efd;
	int child_pid = 0;
	char buffer[256];


	/* open pipes for two one-way communication */
	if(pipe(pipepath1) < 0 || pipe(pipepath2) < 0 || pipe(pipepath3) < 0) {
		strcpy(buffer, "spawn3: failed to pipe with '");
		strncat(buffer,argv[0],100);
		strcat(buffer,"'");
#ifdef RKOERROR
		(void ) rkocpyerror(buffer);
#else
			perror(buffer);
#endif
		return -1;
	}
	/* for clarity only */
	parent_rfd = pipepath1[0];
	child_wfd = pipepath1[1];
	child_rfd = pipepath2[0];
	parent_wfd = pipepath2[1];
	parent_efd = pipepath3[0];
	child_efd = pipepath3[1];

	/* fork process */
	switch(child_pid = fork()) {
	case -1:				/* fork failure */
		strcpy(buffer, "spawn3: failed to fork with '");
		strncat(buffer,argv[0],100);
		strcat(buffer,"'");
#ifdef RKOERROR
		(void ) rkocpyerror(buffer);
#else
			perror(buffer);
#endif
		(void) close(parent_rfd);
		(void) close(child_wfd);
		(void) close(child_rfd);
		(void) close(parent_wfd);
		(void) close(parent_efd);
		(void) close(child_efd);
		return -2;

	case 0:					/* child */
		if (close(parent_rfd) < 0 || close(parent_wfd) < 0
		|| (parent_efd) < 0) {
			strcpy(buffer,
				"spawn3: failed to close pipes with '");
			strncat(buffer,argv[0],100);
			strcat(buffer,"'");
			goto child_failure;
		}
		if (dup2(child_rfd, STDIN_FILENO) < 0) {
			strcpy(buffer,
				"spawn3: child failed to dup2 stdin with '");
			strncat(buffer,argv[0],100);
			strcat(buffer,"'");
			goto child_failure;
		}
		if (dup2(child_wfd, STDOUT_FILENO) < 0) {
			strcpy(buffer,
				"spawn3: child failed to dup2 stdout with '");
			strncat(buffer,argv[0],100);
			strcat(buffer,"'");
			goto child_failure;
		}
		if (dup2(child_efd, STDERR_FILENO) < 0) {
			strcpy(buffer,
				"spawn3: child failed to dup2 stderr with '");
			strncat(buffer,argv[0],100);
			strcat(buffer,"'");
			goto child_failure;
		}
		if (child_rfd != STDIN_FILENO) {
			if(close(child_rfd) < 0) {
				strcpy(buffer,
				"spawn3: child failed to close in pipe to '");
				strncat(buffer,argv[0],100);
				strcat(buffer,"'");
				goto child_failure;
			}
		}
		if (child_wfd != STDOUT_FILENO) {
			if(close(child_wfd) < 0) {
				strcpy(buffer,
				"spawn3: child failed to close out pipe to '");
				strncat(buffer,argv[0],100);
				strcat(buffer,"'");
				goto child_failure;
			}
		}
		if (child_wfd != STDERR_FILENO) {
			if(close(child_efd) < 0) {
				strcpy(buffer,
				"spawn3: child failed to close err pipe to '");
				strncat(buffer,argv[0],100);
				strcat(buffer,"'");
				goto child_failure;
			}
		}
		if (execvp(argv[0],argv) < 0) {
			strcpy(buffer, "spawn3: child failed to execvp with '");
			strncat(buffer,argv[0],100);
			strcat(buffer,"'");
			goto child_failure;
		}
		strcpy(buffer, "spawn3: should not 'get here' with '");
		strncat(buffer,argv[0],100);
		strcat(buffer,"'");
		child_failure: {
#ifdef RKOERROR
		(void ) rkocpyerror(buffer);
#else
			perror(buffer);
#endif
			_exit(1);
		}

	default:				/* parent */
		if (close(child_wfd) < 0 || close(child_rfd) < 0) {
			strcpy(buffer,
			"spawn3: parent failed to close pipes with '");
			strncat(buffer,argv[0],100);
			strcat(buffer,"'");
			goto parent_failure;
		}
		if((*childout = fdopen(parent_rfd, "r")) == (FILE *) NULL) {
			strcpy(buffer,
			"spawn3: parent failed to fdopen readpipe with '");
			strncat(buffer,argv[0],100);
			strcat(buffer,"'");
			goto parent_failure;
		}
		if((*childerr = fdopen(parent_efd, "r")) == (FILE *) NULL) {
			strcpy(buffer,
			"spawn3: parent failed to fdopen errorpipe with '");
			strncat(buffer,argv[0],100);
			strcat(buffer,"'");
			goto parent_failure;
		}
		if((*childin = fdopen(parent_wfd, "w")) == (FILE *) NULL) {
			strcpy(buffer,
			"spawn3: parent failed to fdopen writepipe with '");
			strncat(buffer,argv[0],100);
			strcat(buffer,"'");
			goto parent_failure;
		}
		if(setvbuf(*childout, (char *) NULL, _IOLBF, 0)) {
			strcpy(buffer,
			"spawn3: parent failed to readpipe line buffer with '");
			strncat(buffer,argv[0],100);
			strcat(buffer,"'");
			goto parent_failure;
		}
		/* must not buffer stderr from the child else will lose some */
		if(setvbuf(*childerr, (char *) NULL, _IONBF, 0)) {
			strcpy(buffer,
			"spawn3: parent failed to errpipe line buffer with '");
			strncat(buffer,argv[0],100);
			strcat(buffer,"'");
			goto parent_failure;
		}
		if(setvbuf(*childin, (char *) NULL, _IOLBF, 0)) {
			strcpy(buffer,
			"spawn3: parent failed to writepipe line buffer with '");
			strncat(buffer,argv[0],100);
			strcat(buffer,"'");
			goto parent_failure;
		}
		if (0) parent_failure: {
			(void) close(child_wfd);
			(void) close(parent_rfd);
			(void) close(child_rfd);
			(void) close(parent_wfd);
			(void) close(parent_efd);
			(void) close(child_efd);
#ifdef RKOERROR
		(void ) rkocpyerror(buffer);
#else
			perror(buffer);
#endif
			/* kill the boy... */
			(void) kill(child_pid,SIGKILL);
		}
		/* don't wait for child */
		return child_pid;
	}
}
